File: js/Graph.js
"use strict";
var _=require('underscore');

// window.sandpile = window.sandpile || {};
var sandpile = {};

sandpile.Graph = function(dim) {
	this.var = {};
	this.dim = dim // graph has fixed upper dim for matrix use
};

sandpile.Graph.prototype = {
	addVertex: function(vertex) {
		if (vertex<this.dim*this.dim){this.var[vertex] = [];}
	},
	addEdge: function(v1,v2) { // assumes v1,v2 in the object
		this.var[v1].push(v2);
		if (v1!==v2) {this.var[v2].push(v1);}
	},
	elimVertex: function(vertex) {
		delete this.var[vertex];
		for (var i in Object.keys(this.var)) {
			var elt = Object.keys(this.var)[i];
			var index = this.var[elt].indexOf(vertex);
			if (index!==-1) {this.var[elt].splice(index,1);}
		}
	},
	elimEdge: function(v1,v2) {
		if (this.hasEdge(v1,v2)) {
			var index = this.var[v1].indexOf(v2);
			this.var[v1].splice(index,1);
			index = this.var[v2].indexOf(v1);
			this.var[v2].splice(index,1);
		}		
	},
	hasEdge: function(v1,v2) {
		return this.var[v1].indexOf(v2)!==-1;
	},
	hasVertex: function(vertex) {
		return !_.isUndefined(this.var[vertex]);
	},
	listVertex: function() {
		return Object.keys(this.var);
	},
	listAdjacent: function(vertex) {
		return this.var[vertex];
	}
}
var graph = new sandpile.Graph(10);
graph.addVertex(3);
graph.addVertex(4);
graph.addVertex(2);
graph.addVertex(1);
console.log(graph.listVertex());
graph.addEdge(3,4);
graph.addEdge(3,1);
graph.addEdge(3,3);
console.log(graph.listAdjacent(3));
graph.elimEdge(3,1);
console.log(graph.listAdjacent(3));
graph.elimVertex(3);
console.log(graph.listAdjacent(4));
graph.elimVertex(1);
graph.addVertex(120)
console.log(graph.listVertex());
console.log(graph.listAdjacent(4));

File: java/Graph.java
// Cole Ellison
// Graph Object

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;

public class Graph {
	
	protected HashMap<Integer,HashSet<Integer>> map;

	public Graph() {
		map = new HashMap<Integer,HashSet<Integer>>();
	}

	public void addVertex(int v) {
		map.put(v,new HashSet<Integer>());
	}

	public void addEdge(int v1, int v2) {
		map.get(v1).add(v2);
		map.get(v2).add(v1);
	}

	public void elimVertex(int v) {
		map.remove(v);
		Iterator it = listAdjacent(v).iterator();
	    while (it.hasNext()) {
	    	map.get(it.next()).remove(v);
	    }
	}

	public void elimEdge(int v1, int v2) {
		if (map.containsKey(v1) && map.containsKey(v2)) {
			map.get(v1).remove(v2);
			map.get(v2).remove(v1);
		}
	}

	public boolean hasEdge(int v1, int v2) {
		return map.get(v1).contains(v2);
	}

	public boolean hasVertex(int v) {
		return map.containsKey(v);
	}

	public HashSet listVertex() {
		return new HashSet<Integer>(map.keySet());
	}

	public HashSet listAdjacent(int v) {
		return map.get(v);
	}

	public static void main(String[] args) {
		Graph mygraph = new Graph();
		mygraph.addVertex(3);
		mygraph.addVertex(4);
		mygraph.addVertex(2);
		mygraph.addVertex(1);
		System.out.println(mygraph.listVertex());
		mygraph.addEdge(3,4);
		mygraph.addEdge(4,1);
		mygraph.addEdge(3,3);
		System.out.println(mygraph.listAdjacent(3));
		mygraph.elimEdge(3,3);
		System.out.println(mygraph.listAdjacent(3));
		mygraph.elimVertex(3);
		System.out.println(mygraph.listAdjacent(4));
		mygraph.elimVertex(1);
		System.out.println(mygraph.listVertex());
		System.out.println(mygraph.listAdjacent(4));
	}

}